### PR TITLE
JDK-8318566: Heap walking functions should not use FilteredFieldStream

### PR DESCRIPTION
**Pull request by alexmenkov**
FilteredFieldStream used by heap walking functions to iterate through klass/superclasses/interfaces fields are known to have poor performance (see [JDK-8317692](https://bugs.openjdk.org/browse/JDK-8317692) for details).
Heap walking API implementation is the last user of the klasses.
The fix reworks iteration through klass/superclasses/interfaces fields and drops FilteredFieldStream-related code.
Additionally removed/updated includes of reflectionUtils.hpp.

Testing:
  - tier1..4;
  - test/hotspot/jtreg/vmTestbase/nsk/jvmti (contains tests for different heap walking functions);
  - new test from #17580 (now the test runs several times faster).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318566](https://bugs.openjdk.org/browse/JDK-8318566): Heap walking functions should not use FilteredFieldStream (**Enhancement** - P4)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to [d6ab43b4](https://git.openjdk.org/jdk/pull/17661/files/d6ab43b4363a33401544dd39d2059bd23ff1cf6d)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17661/head:pull/17661` \
`$ git checkout pull/17661`

Update a local copy of the PR: \
`$ git checkout pull/17661` \
`$ git pull https://git.openjdk.org/jdk.git pull/17661/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17661`

View PR using the GUI difftool: \
`$ git pr show -t 17661`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17661.diff">https://git.openjdk.org/jdk/pull/17661.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17661#issuecomment-1920008053)

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 1]
```diff
@@ -424,11 +431,39 @@ ClassFieldMap::~ClassFieldMap() {
...
 void ClassFieldMap::add(int index, char type, int offset) {
   ClassFieldDescriptor* field = new ClassFieldDescriptor(index, type, offset);
   _fields->append(field);
 }
 
+int ClassFieldMap::add_instance_fields(InstanceKlass* ik, int start_index) {
+  // add super fields first
+  InstanceKlass* super_klass = ik->java_super();
+  if (super_klass != nullptr) {
+    start_index += add_instance_fields(super_klass, start_index);
```
Comment (plummercj) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:453]: Does hotspot have any rules against potentially very deep recursion that can overflow the stack?

Comment (alexmenkov) [reply src/hotspot/share/prims/jvmtiTagMap.cpp:453]: I looked at hotspot code and don't see "implementation limits" for level of inheritance, so it looks like a valid concern.
Need to reimplement this without recursion.

Comment (alexmenkov) [reply src/hotspot/share/prims/jvmtiTagMap.cpp]: Fixed.

**Review by plummercj — APPROVED** [review]
Looks good.

**Review by sspitsyn — APPROVED** [review]
Looks good. There is one minor comment though.

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 2]
```diff
@@ -424,55 +427,76 @@ ClassFieldMap::~ClassFieldMap() {
...
+    for (int index = 0; !fld.done(); fld.next(), index++) {
+      // ignore static fields
+      if (fld.access_flags().is_static()) {
+        continue;
+      }
+      field_map->add(start_index + index, fld.signature()->char_at(0), fld.offset());
     }
-    field_map->add(max_field_index - index, fld.signature()->char_at(0), fld.offset());
+    // update total_field_number for superclass
+    total_field_number = start_index;
```
[the remark this answers is not in the document]
Comment (alexmenkov) [reply src/hotspot/share/prims/jvmtiTagMap.cpp:499]: done

**Review by sspitsyn — APPROVED** [review]

**Pull request closed without merging**